node/core/candidate-validation/src/lib.rs

```diff
@@ -709,13 +711,20 @@ trait ValidationBackend {
 			self.validate_candidate(pvf.clone(), exec_timeout, params.encode()).await;
 
 		// Allow limited retries for each kind of error.
+		//
+		// TODO: Should we stop retrying after some time has passed?
+		//       We would need to know if we came from backing or approval.
```

Comment:
@s0me0ne-unkn0wn IIRC you've looked into this, WDYT?

Reply:
Why would a panic resolve itself after retry ? Maybe in case of memory pressure -> OOM  ? But still, if we retry and wait too much time, malicious validators might have enough time to vote valid and the including  block will be finalized - we currently have 2 relay chain blocks delay on finality which should help, but we'd have to be careful here.

Reply:
You mean in the case of backing you want the strict timeout to be applied to both retries cumulatively, not to a single one? Looks like a valid idea. There's no dedicated field to tell from which subsystem the request is coming yet, but you can check for execution timeout kind in the incoming `CandidateValidationMessage::ValidateFromExhaustive` or `CandidateValidationMessage::ValidateFromChainState`, it will be either `PvfExecTimeoutKind::Backing` or `PvfExecTimeoutKind::Approval` so you can tell backing from the other stuff. Sounds a bit hacky probably, but will work.

Reply:
> Why would a panic resolve itself after retry ? Maybe in case of memory pressure -> OOM ?

Maybe, or also just a hardware glitch, spurious OS glitch, etc. We can't really tell why a panic happened, or if it was an issue with the candidate or not.[^1] Retrying lets us be more sure about our vote.

[^1]: Well we do get a stringified error but we can't match on it or anything.

Reply:
Thanks @s0me0ne-unkn0wn! Pushed some changes.

Reply:
But still, if those new guys also retry, they will also no-show and malicious votes come in ?

Reply:
They will then be replaced by even more checkers. One noshow will be replaced by >1 more checkers, which will again be 2/3 honest. Anyhow, this perfect DoS of all honest nodes is certainly the absolute worst case, so yes we have to be careful when to not raise a dispute. In general, if in doubt raise one. 

For hardware faults being the cause, I think we should look into having error checking. E.g. we already suggest/require ECC memory - we should have something similar for disk/db. The node should be able to detect a corrupted db and not just read garbage data and then dispute stuff.

Reply:
CPU hardware faults are extremely rare (I'm closely familiar with a single case, though a really interesting one), and memory faults are not so rare (I came across two cases in my life) but still not something that should be seriously taken into account (that's something that could be handled by governance reverting the slashes on a case-by-case basis), so do I understand correctly we are left with a single really concerning case of OOM? Can we tell the OOM from the other cases? Afair Linux's `oom_kill_task()` sends `SIGKILL` to the process instead of `SIGSEGV`, could we use it to detect the condition?

Reply:
> Can we tell the OOM from the other cases?

Someone (I believe @burdges?) had the idea of setting a flag before each allocation and unsetting it after -- would be easy to do if we had the [allocator wrapper](https://github.com/paritytech/polkadot/issues/6687) that's been discussed. And I agree that we should focus on OOM, since we've seen it happen in the wild.